### ISSUE TITLE
Incorrect compiler error for method with async/non-async delegate overloads

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#2735]
Assuming the following two declarations:

```c#
static void M(Func<Task<int>> f, int i)
{}

static void M(Func<int> f, int i)
{}
```

The code:

```c#
M(() => 42, "");
```

Causes the errors:

> CS0029 Cannot implicitly convert type '`int`' to '`System.Threading.Tasks.Task<int>`'
> CS1662 Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type

While the code:

```c#
M(async () => 42, "");
```

Causes the error:

> CS4010 Cannot convert async lambda expression to delegate type '`Func<int>`'. An async lambda expression may return `void`, `Task` or `Task<T>`, none of which are convertible to '`Func<int>`'.

In both cases, the code is wrong, but the reported error is bogus: the lambda parameter is okay, it's the second parameter that causes the compilation to fail (`string` passed in instead of `int`), so that's what should be reported.

I'm using VS 2015 RC and the latest version of Roslyn from NuGet (1.0.0-rc2) behaves the same. [Full repro code for reference.](https://gist.github.com/svick/08a665153ac9b8567195)

This report is based on dotnet/corefx#1738 by @tucaz.

**Comment by AlekseyTs** [comment]
I am getting the same diagnostics from VS 2013 compiler.

**Comment by jaredpar** [comment]
This appears to be "By Design", or at least not a regression.  As @AlekseyTs noted these are the same errors that we get on VS 2013.

**Comment by svick** [comment]
@jaredpar Okay, it's not a regression, but shouldn't understandable errors be a goal of the compiler? Are you really saying this is the right error message to give in this case (which is how I would understand "By Design")?

**Comment by gchernis** [comment]
Whenever available, a workaround to get a meaningful compiler error could be to comment out the Func<Task<T>> overload and related call-site.  Once that is in place, compiler correctly identifies the second argument's type as the culprit.  

By symmetry, once I resurrect Func<Task<T>> overload and comment out Func<T> overload, compiler provides meaningful guidance as to second argument's type mismatch.
